skyportal/models/observation_plan.py

```diff
@@ -0,0 +1,346 @@
+    )
+
+    instrument_id = sa.Column(
+        sa.ForeignKey('instruments.id', ondelete="CASCADE"),
+        nullable=False,
+        doc='Instrument ID',
+    )
+
+    instrument = relationship(
+        "Instrument",
+        foreign_keys=instrument_id,
+        doc="The Instrument that this observation plan belongs to",
+    )
+
+    dateobs = sa.Column(
+        sa.ForeignKey('gcnevents.dateobs', ondelete="CASCADE"),
+        nullable=False,
+        index=True,
+        doc='UTC event timestamp',
+    )
+
+    plan_name = sa.Column(sa.String, unique=True, doc='Plan name')
+
+    validity_window_start = sa.Column(
+        sa.DateTime,
+        nullable=False,
+        default=datetime.utcnow,
+        doc='Start of validity window',
+    )
+
+    validity_window_end = sa.Column(
+        sa.DateTime,
+        nullable=False,
+        default=datetime.now() + timedelta(1),
+        doc='End of validity window',
+    )
+
+    status = sa.Column(
+        sa.String(),
+        nullable=False,
+        default="pending submission",
+        index=True,
+        doc="The status of the observing plan.",
+    )
+
+    planned_observations = relationship(
+        "PlannedObservation",
+        passive_deletes=True,
+        doc='Planned observations associated with this plan.',
+    )
+
+    @property
+    def start_observation(self):
+        """Time of the first planned observation."""
+        if self.planned_observations:
+            return min(
+                [
+                    planned_observation.obstime
+                    for planned_observation in self.planned_observations
+                ]
+            )
+        else:
+            return None
+
+    @hybrid_property
+    def num_observations(self):
+        """Number of planned observation."""
+        return len(self.planned_observations)
+
+    @num_observations.expression
+    def num_observations(cls):
+        """Number of planned observation."""
+        return cls.planned_observations.count()
+
+    @property
+    def total_time(self):
+        """Total observation time (seconds)."""
+        return sum(_.exposure_time for _ in self.planned_observations)
+
+    @property
+    def tot_time_with_overheads(self):
+        overhead = sum(_.overhead_per_exposure for _ in self.planned_observations)
+        return overhead + self.total_time
+
+    @property
+    def ipix(self):
+        return {
+            i
+            for planned_observation in self.planned_observations
+            if planned_observation.field.ipix is not None
+            for i in planned_observation.field.ipix
+        }
+
+    @property
+    def area(self):
+        nside = Localization.nside
+        return hp.nside2pixarea(nside, degrees=True) * len(self.ipix)
+
+    def get_probability(self, localization):
+        ipix = np.asarray(list(self.ipix))
+        return localization.flat_2d[ipix].sum()
+
+
+class PlannedObservation(Base):
```

Comment:
Whats the difference between `PlannedObservation` and `EventObservationPlan` and can we make that difference clear in the docstrings?

Reply:
Adding some info.